stdlib/public/core/SharedString.swift

```diff
@@ -0,0 +1,104 @@
+  ///
+  /// This initializer does not try to repair ill-formed UTF-8 code unit
+  /// sequences. If any are found, the result of the initializer is `nil`.
+  ///
+  /// - Parameters:
+  ///   - buffer: An `UnsafeBufferPointer` containing the UTF-8 bytes that
+  ///     should be shared with the created `String`.
+  ///   - owner: An optional object that owns the memory referenced by `buffer`
+  ///     and is responsible for deallocating it.
+  public init?(
+    sharingContent buffer: UnsafeBufferPointer<UInt8>,
+    owner: AnyObject?
+  ) {
+    guard let baseAddress = buffer.baseAddress,
+      case .success(let extraInfo) = validateUTF8(buffer)
+    else {
+      return nil
+    }
+    let storage = __SharedStringStorage(
+      immortal: baseAddress,
+      countAndFlags: _StringObject.CountAndFlags(
+        sharedCount: buffer.count,
+        isASCII: extraInfo.isASCII))
+    storage._owner = owner
+    self.init(String(_StringGuts(storage)))
+  }
+
+  /// Creates a `String` whose backing store is shared with the UTF-8 data in
+  /// the given array.
+  ///
+  /// The new `String` instance shares ownership of the array's underlying
+  /// storage, guaranteeing that the `String` remains valid even if the `array`
+  /// is released.
+  ///
+  /// This initializer does not try to repair ill-formed UTF-8 code unit
+  /// sequences. If any are found, the result of the initializer is `nil`.
+  ///
+  /// - Parameter array: An `Array` containing the UTF-8 bytes that should be
+  ///   shared with the created `String`.
+  public init?(sharing array: [UInt8]) {
+    guard case (let owner, let rawPtr?) = array._cPointerArgs() else {
+      self = ""
+      return
+    }
+    let baseAddress = rawPtr.assumingMemoryBound(to: UInt8.self)
+    let buffer = UnsafeBufferPointer(start: baseAddress, count: array.count)
+    self.init(sharingContent: buffer, owner: owner)
+  }
+}
+
+extension Substring {
+  /// Calls the given closure, passing it a `String` whose contents are equal to
+  /// the substring but which shares the substring's storage instead of copying
+  /// it.
+  ///
+  /// The `String` value passed into the closure is only valid durings its
+  /// execution.
+  ///
+  /// - Parameter body: A closure with a `String` parameter that shares the
+  ///   storage of the substring. If `body` has a return value, that value is
+  ///   also used as the return value for the `withSharedString(_:)` method. The
+  ///   argument is valid only for the duration of the closure's execution.
+  /// - Returns: The return value, if any, of the `body` closure parameter.
+  public func withSharedString<Result>(
+    _ body: (String) throws -> Result
+  ) rethrows -> Result {
+    guard _wholeGuts.isFastUTF8 else {
+      // TODO: What's the right way to get the pointer for a non-fast-UTF-8
+      // string here?
+      return try body(String(self))
+    }
+
+    return try _wholeGuts.withFastUTF8 { utf8 in
```

Comment:
This isn't small-string correct, but small substrings will form small strings, so eager copy is even faster than using `__SharedStringStorag`. What we'd want to do is check for a contiguous string with a stable pointer.